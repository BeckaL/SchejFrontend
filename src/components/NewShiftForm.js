import React, { Component } from 'react'
import '../flatpickr-custom.css'
import Flatpickr from 'react-flatpickr'
import axios from 'axios'

export default class NewShiftForm extends Component {
  constructor(props) {
    super(props)

    this.state = {
      currentDate: null,
      startDate: null,
      endDate: null
    }
  }

  componentWillMount() {
    this.setState({
      currentDate: this.formatDate(12),
      startDate: this.formatDate(12),
      endDate: this.formatDate(20)
    })
  }

  formatDate = hours => {
    return new Date(new Date().setHours(new Date().getHours() + hours))
  }

  handleSubmit = event => {
    event.preventDefault()
    this.createNewShiftRequest()
  }

  createNewShiftRequest = () => {
    let stateCopy = this.state
    console.log(stateCopy)
    axios.post(
      `${process.env.REACT_APP_API_URL}/shifts`, {
        user_id: localStorage['id'],
        title: localStorage['name'],
        email: localStorage['email'],
        start_time: stateCopy.startDate.getTime(),
        end_time: stateCopy.endDate.getTime()
      }
    )
    .then(response => {
      this.props.history.push('/')
      this.props.history.push('/shifts')
    })
    .catch(error => {
      console.log(error)
    })
  }

  render() {
    return (
      <div id='new-shift-form'>
        <h3 className='popup-title'>Add a new shift</h3>
        <p className='new-shift-label'>Shift start:</p>
        <Flatpickr data-enable-time
          value={this.state.currentDate}
          onChange={startDate => { this.setState({ startDate: startDate[0] }) }}
          options={{minDate: this.state.currentDate}}
        />
        <br />
        <p className='new-shift-label'>Shift end:</p>
        <Flatpickr data-enable-time
          value={this.state.endDate}
          onChange={endDate => { this.setState({ endDate: endDate[0] }) }}
          options={{minDate: this.state.startDate}}
        />
        <br />
        <form className='new-shift-form' onSubmit={this.handleSubmit}>
          <button id='new-shift-submit' className='custom-button'>Submit</button>
        </form>
      </div>
    )
  }
}
