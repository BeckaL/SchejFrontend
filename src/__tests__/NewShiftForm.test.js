import React from 'react'
import { create } from "react-test-renderer"
import { shallow } from "enzyme"
import Flatpickr from 'react-flatpickr'
import NewShiftForm from '../components/NewShiftForm'
import axios from "axios"
import MockDate from "mockdate"

jest.mock('axios')

describe('new shift form', () => {
  var wrapper

  beforeAll(() => {
    MockDate.set(1545151751163)
    wrapper = shallow(<NewShiftForm history = {[]}/>)
  })

  test("it matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot()
  })

  test("it always renders a div", () => {
    const divs = wrapper.find("div")
    expect(divs.length).toEqual(1)
  })

  test("it always renders 2 Flatpickr calendars", () => {
    const calendars = wrapper.find(Flatpickr)
    expect(calendars.length).toEqual(2)
  })

  test("it renders a submit button", () => {
    const button = wrapper.find("button")
    expect(button.props().children).toEqual("Submit")
  })

  test("makes correct api request when button is clicked", () => {
    axios.post.mockResolvedValue({})
    const form = wrapper.find("form")
    const fakeEvent = { preventDefault () {} }
    form.props().onSubmit(fakeEvent)
    expect(axios.post).toHaveBeenCalledTimes(1)
  })

  afterAll(() => {
    MockDate.reset()
  })
})
