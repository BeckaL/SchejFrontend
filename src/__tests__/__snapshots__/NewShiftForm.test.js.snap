// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new shift form it matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewShiftForm
    history={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="popup-title"
        >
          Add a new shift
        </h3>,
        <p
          className="new-shift-label"
        >
          Shift start:
        </p>,
        <DateTimePicker
          data-enable-time={true}
          onChange={[Function]}
          options={
            Object {
              "minDate": 2018-12-19T04:49:11.163Z,
            }
          }
          value={2018-12-19T04:49:11.163Z}
        />,
        <br />,
        <p
          className="new-shift-label"
        >
          Shift end:
        </p>,
        <DateTimePicker
          data-enable-time={true}
          onChange={[Function]}
          options={
            Object {
              "minDate": 2018-12-19T04:49:11.163Z,
            }
          }
          value={2018-12-19T12:49:11.163Z}
        />,
        <br />,
        <form
          className="new-shift-form"
          onSubmit={[Function]}
        >
          <button
            className="custom-button"
            id="new-shift-submit"
          >
            Submit
          </button>
        </form>,
      ],
      "id": "new-shift-form",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add a new shift",
          "className": "popup-title",
        },
        "ref": null,
        "rendered": "Add a new shift",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Shift start:",
          "className": "new-shift-label",
        },
        "ref": null,
        "rendered": "Shift start:",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data-enable-time": true,
          "onChange": [Function],
          "options": Object {
            "minDate": 2018-12-19T04:49:11.163Z,
          },
          "value": 2018-12-19T04:49:11.163Z,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Shift end:",
          "className": "new-shift-label",
        },
        "ref": null,
        "rendered": "Shift end:",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data-enable-time": true,
          "onChange": [Function],
          "options": Object {
            "minDate": 2018-12-19T04:49:11.163Z,
          },
          "value": 2018-12-19T12:49:11.163Z,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="custom-button"
            id="new-shift-submit"
          >
            Submit
          </button>,
          "className": "new-shift-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "className": "custom-button",
            "id": "new-shift-submit",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="popup-title"
          >
            Add a new shift
          </h3>,
          <p
            className="new-shift-label"
          >
            Shift start:
          </p>,
          <DateTimePicker
            data-enable-time={true}
            onChange={[Function]}
            options={
              Object {
                "minDate": 2018-12-19T04:49:11.163Z,
              }
            }
            value={2018-12-19T04:49:11.163Z}
          />,
          <br />,
          <p
            className="new-shift-label"
          >
            Shift end:
          </p>,
          <DateTimePicker
            data-enable-time={true}
            onChange={[Function]}
            options={
              Object {
                "minDate": 2018-12-19T04:49:11.163Z,
              }
            }
            value={2018-12-19T12:49:11.163Z}
          />,
          <br />,
          <form
            className="new-shift-form"
            onSubmit={[Function]}
          >
            <button
              className="custom-button"
              id="new-shift-submit"
            >
              Submit
            </button>
          </form>,
        ],
        "id": "new-shift-form",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add a new shift",
            "className": "popup-title",
          },
          "ref": null,
          "rendered": "Add a new shift",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Shift start:",
            "className": "new-shift-label",
          },
          "ref": null,
          "rendered": "Shift start:",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data-enable-time": true,
            "onChange": [Function],
            "options": Object {
              "minDate": 2018-12-19T04:49:11.163Z,
            },
            "value": 2018-12-19T04:49:11.163Z,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Shift end:",
            "className": "new-shift-label",
          },
          "ref": null,
          "rendered": "Shift end:",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data-enable-time": true,
            "onChange": [Function],
            "options": Object {
              "minDate": 2018-12-19T04:49:11.163Z,
            },
            "value": 2018-12-19T12:49:11.163Z,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="custom-button"
              id="new-shift-submit"
            >
              Submit
            </button>,
            "className": "new-shift-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "custom-button",
              "id": "new-shift-submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
